Firestore/Swift/Source/Codable/third_party/FirestoreEncoder.swift

```diff
@@ -0,0 +1,616 @@
+
+import Foundation
+import FirebaseFirestore
+
+extension CollectionReference {
+  public func addDocument<T: Encodable>(_ item: T) -> DocumentReference {
+    do {
+      return addDocument(data: try Firestore.Encoder().encode(item))
+    } catch let error {
+      fatalError("Unable to encode data with Firestore encoder: \(error)")
+    }
+  }
+
+  public func addDocument<T: Encodable>(_ item: T, _ completion: ((Error?) -> Void)?) -> DocumentReference {
+    do {
+      let encoded = try Firestore.Encoder().encode(item)
+      return addDocument(data: encoded, completion: completion)
+    } catch let error {
+      Firestore.firestore().settings.dispatchQueue.sync {
+        completion!(error)
+      }
+      return document() // Is there something better to return after the error?
+    }
+  }
+}
+
+extension DocumentReference {
+  public func setData<T: Encodable>(_ value: T) {
+    do {
+      setData(try Firestore.Encoder().encode(value))
+    } catch let error {
+      fatalError("Unable to encode data with Firestore encoder: \(error)")
+    }
+  }
+
+  public func setData<T: Encodable>(_ value: T, _ completion: ((Error?) -> Void)?) {
+    do {
+      let encoded = try Firestore.Encoder().encode(value)
+      setData(encoded, completion: completion)
+    } catch let error {
+      Firestore.firestore().settings.dispatchQueue.sync {
+        completion!(error)
+      }
+    }
+  }
+}
+
+extension Transaction {
+  public func setData<T: Encodable>(_ value: T, forDocument: DocumentReference) {
+    do {
+      setData(try Firestore.Encoder().encode(value), forDocument: forDocument)
+    } catch let error {
+      fatalError("Unable to encode data with Firestore encoder: \(error)")
+    }
+  }
+}
+
+extension WriteBatch {
+  public func setData<T: Encodable>(_ value: T, forDocument: DocumentReference) {
+    do {
+      setData(try Firestore.Encoder().encode(value), forDocument: forDocument)
+    } catch let error {
+      fatalError("Unable to encode data with Firestore encoder: \(error)")
+    }
+  }
+}
+
+extension Firestore {
+  struct Encoder {
+    func encode<T: Encodable>(_ value: T) throws -> [String: Any] {
+      guard let topLevel = try _FirestoreEncoder().box_(value) else {
+        throw EncodingError.invalidValue(value,
+                                         EncodingError.Context(codingPath: [],
+                                                               debugDescription: "Top-level \(T.self) did not encode any values."))
+      }
+
+      // This is O(n) check. Consider refactoring box_ to return [String: Any].
+      guard let dict = topLevel as? [String: Any] else {
+        throw EncodingError.invalidValue(value,
+                                         EncodingError.Context(codingPath: [],
+                                                               debugDescription: "Top-level \(T.self) encoded not as dictionary."))
+      }
+
+      return dict
+    }
+  }
+}
+
+fileprivate class _FirestoreEncoder: Encoder {
```

Comment:
I can't see what you're changing in here. I'd like us to do a separate PR for moving the FirestoreEncoder into third_party so that it's more obvious we're changing.

Reply:
This is a new file to master and mostly unchanged from the original PR. Perhaps, best to look at commit by commit history?